src/sentry/auth/idpmigration.py

```diff
@@ -14,60 +15,79 @@
-
-    :param user: the user profile to link
-    :param org: the organization whose SSO provider is being used
-    :param email: the email address associated with the SSO identity
-    :param identity_id: the SSO identity id
-    """
-    cluster = redis.clusters.get("default").get_local_client_for_key(_REDIS_KEY)
-    member_id = OrganizationMember.objects.get(organization=org, user=user).id
-
-    verification_code = get_random_string(32, string.ascii_letters + string.digits)
-    verification_key = f"auth:one-time-key:{verification_code}"
-    verification_value = {
-        "user_id": user.id,
-        "email": email,
-        "member_id": member_id,
-        "identity_id": identity_id,
-    }
-    cluster.hmset(verification_key, verification_value)
-    cluster.expire(verification_key, int(_TTL.total_seconds()))
-
-    send_confirm_email(user, email, verification_code)
-
-    return verification_code
+    return AccountConfirmLink(user, org, provider_name, email, identity_id).send_one_time_link()
+
+
+@dataclass
+class AccountConfirmLink:
+    user: User
+    organization: Organization
+    provider_name: str
+    email: str
+    identity_id: str
+
+    def send_confirm_email(self, verification_key: str) -> None:
+        context = {
+            "user": self.user,
+            "organization": self.organization.name,
+            "provider": self.provider_name,
+            "url": absolute_uri(
+                reverse(
+                    "sentry-idp-email-verification",
+                    args=[verification_key],
+                )
+            ),
+            "email": self.email,
+            "verification_key": verification_key,
+        }
+        msg = MessageBuilder(
+            subject="{}Confirm Account".format(options.get("mail.subject-prefix")),
+            template="sentry/emails/idp_verification_email.txt",
+            html_template="sentry/emails/idp_verification_email.html",
+            type="user.confirm_email",
+            context=context,
+        )
+        msg.send_async([self.email])
+
+    def send_one_time_link(self) -> str:
+        """Store and email a verification key for IdP migration.
+
+        Create a one-time verification key for a user whose SSO identity
+        has been deleted, presumably because the parent organization has
+        switched identity providers. Store the key in Redis and send it
+        in an email to the associated address.
+
+        :param auth_handler: the object handling the login attempt
+        :param user: the user profile to link
+        :param email: the email address associated with the SSO identity
+        :param identity_id: the SSO identity id
+        """
+        cluster = redis.clusters.get("default").get_local_client_for_key(_REDIS_KEY)
+        member_id = OrganizationMember.objects.get(
+            organization=self.organization, user=self.user
+        ).id
+
+        verification_code = get_random_string(32, string.ascii_letters + string.digits)
```

Comment:
would it be worth moving `verification_code` to `init` (or `post_init` i think since its a dataclass), and also perhaps moving the redis functionality into its own class / function?

Reply:
yup for sure!

Reply:
moving the redis functionaility to its own function is already done in another pr

Reply:
Yep, I'll be waiting to merge this until one or two of those dev branches are finished. I'll have to shake out a few merge conflicts but it won't be a problem.